src/writers.rs

```diff
@@ -0,0 +1,381 @@
+    #[new]
+    pub fn new() -> Self {
+        Self {
+            buff: Arc::new(Mutex::new(Vec::new())),
+            element_count: 0,
+        }
+    }
+
+    pub fn len(&self) -> usize {
+        self.buff.lock().unwrap().len()
+    }
+
+    pub fn set_element_count(&mut self, element_count: u16) {
+        self.element_count = element_count;
+    }
+}
+
+impl SerializationBuffer {
+    pub fn get_serialized_values(&self) -> SerializedValues {
+        let buff = self.buff.lock().unwrap();
+        SerializedValues {
+            serialized_values: buff.clone(),
+            element_count: self.element_count,
+        }
+    }
+}
+
+impl Default for SerializationBuffer {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[pyclass]
+pub struct PyRowWriter {
+    value_count: usize,
+    buff: Arc<Mutex<Vec<u8>>>,
+}
+
+#[pymethods]
+impl PyRowWriter {
+    #[new]
+    pub fn new(buffer: PyRef<'_, SerializationBuffer>) -> Self {
+        Self {
+            value_count: 0,
+            buff: Arc::clone(&buffer.buff),
+        }
+    }
+
+    pub fn make_cell_writer(&mut self) -> PyCellWriter {
+        self.value_count += 1;
+        PyCellWriter {
+            buff: Arc::clone(&self.buff),
+            write_size: true,
+        }
+    }
+
+    pub fn value_count(&self) -> usize {
+        self.value_count
+    }
+}
+
+#[pyclass]
+pub struct PyCellWriter {
+    buff: Arc<Mutex<Vec<u8>>>,
+    write_size: bool,
+}
+
+#[pymethods]
+impl PyCellWriter {
+    #[new]
+    pub fn new(buffer: PyRef<'_, SerializationBuffer>) -> Self {
+        Self {
+            buff: Arc::clone(&buffer.buff),
+            write_size: true,
+        }
+    }
+
+    #[staticmethod]
+    pub fn new_without_size(buffer: PyRef<'_, SerializationBuffer>) -> Self {
+        Self {
+            buff: Arc::clone(&buffer.buff),
+            write_size: false,
+        }
+    }
+
+    pub fn set_value(&mut self, value: &[u8]) -> Result<(), PySerializationError> {
+        let mut buff = self.buff.lock().unwrap();
+        let mut_writer = if self.write_size {
+            CellWriter::new(&mut buff)
+        } else {
+            CellWriter::new_without_size(&mut buff)
+        };
+
+        mut_writer.set_value(value)?;
+        Ok(())
+    }
+
+    pub fn set_null(&mut self) {
+        let mut buff = self.buff.lock().unwrap();
+        let mut_writer = if self.write_size {
+            CellWriter::new(&mut buff)
+        } else {
+            CellWriter::new_without_size(&mut buff)
+        };
+        mut_writer.set_null();
+    }
+
+    pub fn set_unset(&mut self) {
+        let mut buff = self.buff.lock().unwrap();
+
+        let mut_writer = if self.write_size {
+            CellWriter::new(&mut buff)
+        } else {
+            CellWriter::new_without_size(&mut buff)
+        };
+        mut_writer.set_unset();
+    }
+
+    pub fn into_value_builder(&self) -> PyCellValueBuilder {
+        PyCellValueBuilder::new(self.buff.clone(), self.write_size)
+    }
```

Comment:
:wrench: `into` is not appropriate for `borrowed -> owned` conversions. See [the naming conventions](https://rust-lang.github.io/api-guidelines/naming.html#ad-hoc-conversions-follow-as_-to_-into_-conventions-c-conv).